src/projection.jl

```diff
@@ -134,14 +134,32 @@ ProjectTo(::Bool) = ProjectTo{NoTangent}()  # same projector as ProjectTo(::Abst
 ProjectTo(::Real) = ProjectTo{Real}()
 ProjectTo(::Complex) = ProjectTo{Complex}()
 ProjectTo(::Number) = ProjectTo{Number}()
+
+ProjectTo(x::Integer) = ProjectTo(float(x))
+ProjectTo(x::Complex{<:Integer}) = ProjectTo(float(x))
+
+# Preserve low-precision floats as accidental promotion is a common performance bug
 for T in (Float16, Float32, Float64, ComplexF16, ComplexF32, ComplexF64)
-    # Preserve low-precision floats as accidental promotion is a common perforance bug
     @eval ProjectTo(::$T) = ProjectTo{$T}()
 end
-ProjectTo(x::Integer) = ProjectTo(float(x))
-ProjectTo(x::Complex{<:Integer}) = ProjectTo(float(x))
-(::ProjectTo{T})(dx::Number) where {T<:Number} = convert(T, dx)
-(::ProjectTo{T})(dx::Number) where {T<:Real} = convert(T, real(dx))
+
+# In these cases we can just `convert` as we know we are dealing with plain and simple types
+(::ProjectTo{T})(dx::AbstractFloat) where T<:AbstractFloat = convert(T, dx)
+(::ProjectTo{T})(dx::Integer) where T<:AbstractFloat = convert(T, dx)  #needed to avoid ambiguity
+# simple Complex{<:AbstractFloat}} cases
+(::ProjectTo{T})(dx::Complex{<:AbstractFloat}) where {T<:Complex{<:AbstractFloat}} = convert(T, dx)
+(::ProjectTo{T})(dx::AbstractFloat) where {T<:Complex{<:AbstractFloat}} = convert(T, dx)
+(::ProjectTo{T})(dx::Complex{<:Integer}) where {T<:Complex{<:AbstractFloat}} = convert(T, dx)
+(::ProjectTo{T})(dx::Integer) where {T<:Complex{<:AbstractFloat}} = convert(T, dx)
+
+# Other numbers, including e.g. ForwardDiff.Dual and Symbolics.Sym, should pass through.
+# We assume (lacking evidence to the contrary) that it is the right subspace of numebers
+# The (::ProjectTo{T})(::T) method doesn't work because we are allowing a different
+# Number type that might not be a subtype of the `project_type`.
```

Comment:
Is this true? `project_type` should be Real, Complex, or Number. I get super-confused by dispatch with type parameters on LHS like `{T}(...`. But not important really.

Reply:
for e.g. `Float32` or `ComplexF64` etc it is not the case that it is `Real`, `Complex`, or `Number`.
I could add those in as alternative special cases instead of this.

Reply:
I would not be shocked if this come backs to bite us, but we can remove it when that happens

Reply:
Oh right, I wasn't thinking clearly, that's exactly the use case.